File: 킬링캠프/3주 차 - BFS, DFS/practice/Number of Islands/Java/bfs/Solution.java
import java.util.*;

class Solution {

    private static final int[][] DIR = new int[][]{{-1, 0}, {0, 1}, {0, -1}, {1, 0}}; // 북, 동, 서, 남
    private static final char LAND = '1';
    private static final char WATER = '0';

    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int result = 0;
        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid[row].length; col++) {
                if (grid[row][col] == WATER || visited[row][col]) {
                    continue;
                }

                bfs(grid, visited, new Node(row, col));
                result++;
            }
        }

        return result;
    }

    private void bfs(char[][] grid, boolean[][] visited, Node start) {
        Deque<Node> dq = new ArrayDeque<>();
        dq.offer(start);
        visited[start.row][start.col] = true;

        while (!dq.isEmpty()) {
            Node cur = dq.poll();
            for (int i = 0; i < 4; i++) {
                Node next = new Node(cur.row + DIR[i][0], cur.col + DIR[i][1]);
                if (!isSafe(grid, next) || visited[next.row][next.col]) {
                    continue;
                }

                dq.offer(next);
                visited[next.row][next.col] = true;
            }
        }
    }

    private boolean isSafe(char[][] grid, Node node) {
        int row = node.row;
        int col = node.col;

        return (row >= 0 && row < grid.length && col >= 0 && col < grid[row].length && grid[row][col] == LAND);
    }
}

class Node {
    int row;
    int col;

    Node(int row, int col) {
        this.row = row;
        this.col = col;
    }
}

File: 킬링캠프/3주 차 - BFS, DFS/practice/Number of Islands/bfs/solution.py
from collections import deque

class Solution:

    DIR = [[-1, 0], [0, 1], [0, -1], [1, 0]]

    def numIslands(self, grid: List[List[str]]) -> int:
        answer = 0
        visited = [[False for _ in range(len(grid[0]))] for _ in range(len(grid))]

        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "0" or visited[row][col]:
                    continue

                self.bfs(grid, visited, row, col)
                answer +=1
        
        return answer

    def bfs(self, grid, visited, start_row, start_col):
        dq = deque()
        dq.append([start_row, start_col])
        visited[start_row][start_col] = True

        while dq:
            current = dq.popleft()
            for dr, dc in self.DIR:
                next_row = current[0] + dr
                next_col = current[1] + dc

                if not self.is_safe(next_row, next_col, grid, visited):
                    continue

                visited[next_row][next_col] = True
                dq.append([next_row, next_col])
    
    def is_safe(self, row, col, grid, visited):
        return row >= 0 and row < len(grid) and col >= 0 and col < len(grid[row]) and grid[row][col] == "1" and not visited[row][col]
